modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/cluster/programs/Joining.scala

```diff
@@ -190,6 +77,50 @@ sealed abstract class Joining[F[_]: Async: GenUUID: SecurityProvider: KryoSerial
     } yield ()
   }.onError(err => logger.error(err)(s"Error during join attempt by ${joinRequest.registrationRequest.id.show}"))
 
+  private def validateJoinConditions(): F[Unit] =
+    for {
+      nodeState <- nodeStorage.getNodeState
+      canJoinCluster <- nodeStorage.canJoinCluster
+      _ <- Applicative[F].unlessA(canJoinCluster)(NodeStateDoesNotAllowForJoining(nodeState).raiseError[F, Unit])
+    } yield ()
+
+  private def joinAndDiscover(context: P2PContext): EitherT[F, Throwable, Set[P2PContext]] =
+    EitherT(twoWayHandshake(context, none).attempt).flatMapF { peer =>
+      clusterStorage.getPeer(peer.id).map {
+        case Some(p) => p.asRight[Throwable]
+        case None    => new Exception(s"Peer ${peer.show} was not found").asLeft[Peer]
+      }
+    }.flatMapF { peer =>
+      peerDiscovery
+        .discoverFrom(peer)
+        .attempt
+    }
+      .map(_.map(toP2PContext))
+
+  private def joinAll(toJoin: PeerToJoin): F[Unit] = {
+    type Agg = (Set[P2PContext], Set[P2PContext])
+    type Res = Unit
+
+    (Set(toJoin): Set[P2PContext], Set.empty[P2PContext]).tailRecM {
+      case (joining, _) if joining.isEmpty =>
+        ().asRight[Agg].pure
+      case (joining, attempted) =>
+        for {
+          discovered <- joining.toList.parTraverse { peer =>
```

Comment:
I think it will always default to an empty collection of discovered peers (as traverse fails the whole collection traversing) on a single error and I don't think you are successfully handling the errors coming from `joinAndDisccover` function right now.

Reply:
Changed the signature to

```scala
  private def joinAndDiscover(toJoin: P2PContext): EitherT[F, Throwable, Set[P2PContext]]
```

So, the `parTraverse` will go through the entire list, accumulating errors.